eclipse-theia/theia
# reload plugins on reconnect

#### What it does


- [x] fix #5025: reload plugins on reconnect
  - [x] unload plugin static contributions
    - [x] only if a plugin was undeployed
    - [x] unload languages
    - [x] unload tasks
    - [x] unload commands
    - [x] unload menus
    - [x] unload view containers
    - [x] unload views
    - [x] unload keymaps
    - [x] unload schemas
    - [x] unload snippets
    - [x] unload shared css styles
  - [x] unload plugin code and dynamic contributions
    - [x] unload rpc (dispose internal state and does not allow to add new proxies, services and so on)
    - [x] unload command registry (dispose command and handlers)
    - [x] unload quick open (hide quick open and dispose its listeners)
    - [x] unload workspace (dispose workspace and fs watch listeners, content providers) 
      - [x] install only one fs watcher for all plugins instead of for each plugin
    - [x] unload dialogs (no listeners or state to dispose)
    - [x] unload message registry (no listeners or state to dispose)
    - [x] unload preference registry (dispose preference service listeners)
    - [x] unload documents (dispose editor and model listeners)
    - [x] unload editors (remove editor decorations)
    - [x] unload status bar registry (remove status bar entries)
    - [x] unload env (no listeners or state to dispose)
    - [x] unload notification (cancel all running progress reporting)
    - [x] unload terminal (dispose terminal service and terminal listeners)
    - [x] unload tree view (unregister tree data providers and remove tree view listeners)
      - [x] preserve tree views on reconnect
    - [x] unload output (not listeners, output channels are preserved)
    - [x] unload languages (language features gets disposed, but diagnostics are preserved)
    - [x] unload webview (dispose webviews and its serializers)
    - [x] unload storage (no listeners or state to dispose, but path service is rewritten to avoid forever pending promises and accepting workspace URI as an arg for each request instead of caching a random one)
    - [x] unload connection (remove connections)
    - [x] unload tasks (unregister task providers, preserve running tasks)
    - [x] unload languages contribution (stop them)
    - [x] unload debug (dispose debug listeners and remove debug contributions)
      - debug sessions are not preserved, we have to implement: https://github.com/eclipse-theia/theia/issues/3361
    - [x] unload fs (dispose fs providers)
    - [x] unload scm (dispose scm listeners and remove contributed repositories)
    - [x] unload explorer decorations (dispose providers)
    - [x] unload window (remove event listeners)
    - [x] unload clipboard (no listeners or state to dispose)
    - [x] unload shared css styles
- [x] fix #5829: ensure that each plugin is loaded only once
- [x] revert https://github.com/theia-ide/theia/pull/5257: it was bogus to introduce in the first place
- [x] fix #6186: start only one plugin host process for the same host instead of a new for each load as before
- [x] update the changelog with a breaking change
  - [x] clarify how to treat custom main services, see https://github.com/eclipse-theia/theia/pull/6159#issuecomment-532993749
- [x] fix #4590: plugin debug sessions are closed on reconnection for now, track https://github.com/eclipse-theia/theia/issues/4591 and https://github.com/eclipse-theia/theia/issues/3361 for improvements

Out of scope:
- handling extension dependencies on stop
- UI to uninstall extensions

#### How to test


- [x] reconnect because of a connection failure
  - no reproducible with Chrome since it cannot throttle web socket connections
  - throttle with the system proxy instead, for instance with https://www.charlesproxy.com/
- [x] reconnect on the server restart
- [x] test everything with multiple windows for the same backend

Hints:
- Take memory snapshots when a connection is gone to check that there is no main services are left.

Tested with following VS Code extensions:
- [x] trees - [tree-view-sample](https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample)
- [x] languages - [ms-vscode.Go](https://marketplace.visualstudio.com/items?itemName=ms-vscode.Go)
- [x] debugging - [node-debug](https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug) and [node-debug2](https://marketplace.visualstudio.com/items?itemName=ms-vscode.node-debug2)
- [x] scm - [built-in git](https://registry.npmjs.org/@theia/vscode-builtin-git/-/vscode-builtin-git-0.2.1.tgz)
- [x] webviews - [webview-sample](https://github.com/microsoft/vscode-extension-samples/tree/master/webview-sample)
  - make sure to use [this commit](https://github.com/microsoft/vscode-extension-samples/commit/104c10a3f20c347d1a8e4c2cedc01391e6f677dc), never versions use `asWebviewUri` api
- [x] quick pick/input - [quickinput-sample](https://github.com/microsoft/vscode-extension-samples/tree/master/quickinput-sample)

##### Reconnecting tree views
![reconnect_tree_views](https://user-images.githubusercontent.com/3082655/65027015-0f141c80-d953-11e9-8be1-48b353f518e3.gif)

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
